grid2op/multi_agent/multiAgentEnv.py

```diff
@@ -0,0 +1,1070 @@
+        #        #TODO following parameters
+        #        with_forecast=True,
+        #        kwargs_observation=None,
+        #    )
+    
+    def _build_action_spaces(self):
+        """Build action spaces from given domains for each agent
+        The action class is the same as 
+        """
+        for agent in self.agents: 
+            _cls_agent_action_space = ActionSpace.init_grid(gridobj=self._subgrids_cls['action'][agent], extra_name=agent)
+            self.action_spaces[agent] = _cls_agent_action_space(
+                gridobj=self._subgrids_cls['action'][agent],
+                actionClass=self._cent_env._actionClass,
+                legal_action=self._cent_env._game_rules.legal_action,
+            )
+    
+    
+    def _update_observations(self, observation : BaseObservation):
+        """Update agents' observations from the global observation given by the self._cent_env
+
+        Args:
+            observation (BaseObservation): _description_
+        """
+        if self.global_obs:
+            for agent in self.agents:
+                self.observations[agent] = observation.copy()
+        else:
+            #TODO
+            raise NotImplementedError("Local observations are not available yet !")
+    
+    def _verify_domains(self, domains : MADict) :
+        """It verifies if substation ids are valid
+
+        Args:
+            domains (dict): 
+                - key : agents' names
+                - value : list of substation ids
+        """
+        sum_subs = 0
+        for agent in domains.keys():
+            if not (isinstance(domains[agent], list)
+                    or isinstance(domains[agent], set)
+                    or isinstance(domains[agent], np.ndarray)
+            ):
+                raise DomainException("Agent id {agent} : The domain must be a list or a set of substation indices")
+            
+            if len(domains[agent]) == 0:
+                raise DomainException(f"Agent id {agent} : The domain is empty !")
+            
+            for sub_id in domains[agent] : 
+                if not (isinstance(sub_id, int) 
+                        or isinstance(sub_id, dt_int) 
+                        or isinstance(sub_id, np.int32) 
+                        or isinstance(sub_id, np.int64)
+                ):
+                    raise DomainException(f"Agent id {agent} : The id must be of type int. Type {type(sub_id)} is not valid")
+                if sub_id < 0 or sub_id > len(self._cent_env.name_sub) :
+                    raise DomainException(f"Agent id {agent} : The substation's id must be between 0 and {len(self._cent_env.name_sub)-1}, but {sub_id} has been given")
+            sum_subs += len(domains[agent])
+
+        if sum_subs != self._cent_env.n_sub:
+            raise DomainException(f"The sum of sub id lists' length must be equal to _cent_env.n_sub = {self._cent_env.n_sub} but is {sum_subs}")
+    
+    
+    def observation_space(self, agent):
```

Comment:
be specific here:

does it expects and agent id (0, 1, etc.) or an agent name ?

Reply:
Agents are identified by their names/id. It's the same and it's a string. They are not identified by an integer.
I'll add ": str" to clarify that.